src/Utils/Traceroute.ts

```diff
@@ -0,0 +1,236 @@
+  outStream: process.stdout
+};
+
+export class Traceroute extends EventEmitter {
+  private readonly icmpSocket = raw.createSocket({
+    protocol: raw.Protocol.ICMP
+  });
+  private readonly resolver = new dns.Resolver();
+  private readonly options: Options;
+  private port = 33433;
+  private ttl = 1;
+  private startTime?: [number, number];
+  private probes = 0;
+  private timeout?: NodeJS.Timeout;
+  private previousIP?: string;
+  private udpSocket?: dgram.Socket;
+  private destinationHostname: string;
+
+  constructor(
+    private destinationIp: string,
+    userOptions: Partial<Options> = {}
+  ) {
+    super();
+    this.options = { ...defaultOptions, ...userOptions };
+    this.destinationHostname = this.destinationIp;
+
+    this.icmpSocket.on('message', async (buffer: Buffer, ip: string) => {
+      const port = this.udpSocket
+        ? buffer.readUInt16BE(50)
+        : buffer.readUInt16BE(buffer.length - 2);
+      if (port === this.port) {
+        const hostName = await this.getHostName(ip);
+        this.handleReply(ip, hostName);
+      }
+    });
+  }
+
+  public async start(): Promise<void> {
+    if (!isIP(this.destinationIp)) {
+      this.destinationIp = (
+        await this.resolver.resolve(this.destinationHostname, 'A')
+      )[0];
+    }
+
+    process.stdout.write(
+      `traceroute to ${this.destinationHostname} (${this.destinationIp}), ${this.options.maximumHops} hops max, ${this.options.packetSize} byte packets`
+    );
+
+    if (this.options.protocol === raw.Protocol.UDP) {
+      this.udpSocket = dgram.createSocket('udp4');
+      this.udpSocket.bind(() => this.sendPacket());
+    } else {
+      setImmediate(this.sendPacket.bind(this));
+    }
+  }
+
+  public close(): void {
+    if (this.udpSocket) {
+      this.udpSocket.close();
+    }
+    this.icmpSocket.close();
+  }
+
+  private async getHostName(ip: string): Promise<string | undefined> {
+    if (!this.options.reverseLookup) {
+      return;
+    }
+
+    try {
+      return (await this.resolver.reverse(ip))[0];
+    } catch (_e) {
+      return;
+    }
+  }
+
+  private sendPacket(): void {
+    this.probes++;
+
+    if (this.probes > this.options.amountProbes) {
+      this.probes = 1;
+      this.ttl++;
+    }
+
+    const buffer = this.createPingRequest(
+      0,
+      0,
+      ++this.port,
+      this.options.packetSize
+    );
+
+    if (this.udpSocket) {
+      this.udpSocket.setTTL(this.ttl);
+      this.udpSocket.send(
+        buffer,
+        0,
+        buffer.length,
+        this.port,
+        this.destinationIp,
+        this.afterSend.bind(this)
+      );
+    } else {
+      this.icmpSocket.setOption(
+        raw.SocketLevel.IPPROTO_IP,
+        raw.SocketOption.IP_TTL,
+        this.ttl
+      );
+      this.icmpSocket.send(
+        buffer,
+        0,
+        buffer.length,
+        this.destinationIp,
+        this.afterSend.bind(this)
+      );
+    }
+  }
+
+  private afterSend(error: Error | null, _bytes: number) {
```

Comment:
Remove unused argument

```suggestion
  private afterSend(error: Error | null) {
```